warp10/src/main/java/io/warp10/script/functions/DURATIONBUCKETIZE.java

```diff
@@ -0,0 +1,420 @@
+    }
+
+    if (!(params.get(params.size() - 3) instanceof Long) || !(params.get(params.size() - 2) instanceof String) || !(params.get(params.size() - 1) instanceof Long)) {
+      throw new WarpScriptException(getName() + " expects lastbucket, bucketduration, bucketcount (and optionally timezone) as last parameters.");
+    }
+
+    Object bucketizer = params.get(params.size() - 4);
+    long lastbucket = (long) params.get(params.size() - 3);
+    String bucketduration = (String) params.get(params.size() - 2);
+    long bucketcount = (long) params.get(params.size() - 1);
+
+    //
+    // Check that lastbucket is not 0
+    //
+
+    if (0 == lastbucket) {
+      throw new WarpScriptException(getName() + " does not allow lastbucket to be 0. It must be specified.");
+    }
+
+    //
+    // Check that bucketcount is not negative or null and not over maxbuckets
+    //
+
+    if (bucketcount < 0) {
+      throw new WarpScriptException(getName() + " expects a positive bucketcount.");
+    }
+
+    long maxbuckets = (long) stack.getAttribute(WarpScriptStack.ATTRIBUTE_MAX_BUCKETS);
+    if (bucketcount > maxbuckets) {
+      throw new WarpScriptException(getName() + " error: bucket count (" + bucketcount + ") would exceed maximum value of " + maxbuckets);
+    }
+
+    //
+    // Check that input gts are not already duration-bucketized
+    //
+
+    for (GeoTimeSerie gts : series) {
+      Map<String, String> attributes = gts.getMetadata().getAttributes();
+      if (attributes.get(DURATION_ATTRIBUTE_KEY) != null || attributes.get(OFFSET_ATTRIBUTE_KEY) != null || attributes.get(TIMEZONE_ATTRIBUTE_KEY) != null) {
+        throw new WarpScriptException(getName() + " expects GTS for which the attributes " + DURATION_ATTRIBUTE_KEY + ", " + OFFSET_ATTRIBUTE_KEY + " and " + TIMEZONE_ATTRIBUTE_KEY + " are not set. If an input GTS is supposed to be already duration-bucketized, duration-unbucketize it first before applying a new duration-bucketization.");
+      }
+    }
+
+    //
+    // Check nullity of bucketizer
+    //
+
+    if (null == bucketizer) {
+      throw new WarpScriptException(getName() + " expects a non null bucketizer.");
+    }
+
+    //
+    // Convert duration to joda.time.Period
+    //
+
+    if ('P' != bucketduration.charAt(0)) {
+      throw new WarpScriptException(getName() + " expects that the bucketduration is in ISO8601 duration format.");
+    }
+
+    ADDDURATION.ReadWritablePeriodWithSubSecondOffset bucketperiod;
+    try {
+      bucketperiod = ADDDURATION.durationToPeriod(bucketduration);
+    } catch (WarpScriptException wse) {
+      throw new WarpScriptException(getName() + " encountered an exception: " + wse.getMessage());
+    }
+
+    //
+    // Check that the bucketduration is positive
+    //
+
+    long averageSpan = bucketperiod.getPeriod().toPeriod().toDurationFrom(new Instant()).getMillis() * Constants.TIME_UNITS_PER_MS + bucketperiod.getOffset();
```

Comment:
Are you sure about creating a new Instant each time?
Could it be better to chose a fixed instant away from known special dates for instance to avoid leap-second, February and daylight saving changes?

Reply:
Leap seconds are not supported by Unix timestamp. We don't use any particular timezone here so it is UTC which does not has daylight saving time. There can be a point with February but 1970 is not a leap year so that would only applies for durations > 2 years and that's not very impactful on our index hint.